proposals/0000-unicode-for-string-processing.md

```diff
@@ -0,0 +1,1321 @@
+  /// defined behavior of `\w`.
+  public static var simple: Self { get }
+
+  /// A word boundary algorithm that implements the "default word boundary"
+  /// Unicode recommendation.
+  ///
+  /// Default word boundaries use a Unicode algorithm that handles some cases
+  /// better than simple word boundaries, such as words with internal
+  /// punctuation, changes in script, and Emoji.
+  public static var default: Self { get }
+}
+```
+
+#### Matching semantic level
+
+FIXME: Add intro paragraph
+
+These specific levels of matching, and the options to switch between them, are unique to Swift, but not unprecedented in other regular expression engines. Several engines, including Perl, Java, and ICU-based engines like `NSRegularExpression`, support the `\X` metacharacter for matching a grapheme cluster within otherwise Unicode scalar semantic matching. Rust has a related concept in its [`regex::bytes` type][regexbytes], which matches over abitrary bytes by default but allows switching into Unicode mode for segments of the regular expression.
+
+When matching with *grapheme cluster semantics* (the default), metacharacters like `.` and `\w`, custom character classes, and character class instances like `.any` match a grapheme cluster, corresponding with the default string representation. In addition, matching with grapheme cluster semantics compares characters using their canonical representation, corresponding with the way comparing strings for equality works.
+
+When matching with *Unicode scalar semantics*, metacharacters and character classes match a single Unicode scalar value, even if that scalar comprises part of a grapheme cluster.
+
+These semantic levels lead to different results when working with strings that have characters made up of multiple Unicode scalar values, such as Emoji or decomposed characters. In the following example, `queRegex` matches any 3-character string that begins with `"q"`.
+
+```swift
+let composed = "qué"
+let decomposed = "que\u{301}"
+	
+let queRegex = /^q..$/
+	
+print(composed.contains(queRegex))
+// Prints "true"
+print(decomposed.contains(queRegex))
+// Prints "true"
+```
+
+When using Unicode scalar semantics, however, the regex only matches the composed version of the string, because each `.` matches a single Unicode scalar value.
+
+```swift
+let queRegexScalar = queRegex.matchingSemantics(.unicodeScalar)
+print(composed.contains(queRegexScalar))
+// Prints "true"
+print(decomposed.contains(queRegexScalar))
+// Prints "false"
+```
+
+With grapheme cluster semantics, a grapheme cluster boundary is naturally enforced at the start and end of the match and every capture group. Matching with Unicode scalar semantics, on the other hand, can yield string indices that aren't aligned to character boundaries. Take care when using indices that aren't aligned with grapheme cluster boundaries, as they may have to be rounded to a boundary if used in a `String` instance.
```

Comment:
The first sentence packs a lot. It seems like its own paragraph if not section to unpack what it implies. Also, what's special about captures that isn't special about anything else in a regex? Why isn't there a boundary around an alternation, character class, etc?

Reply:
Maybe a sub-section about the outputs of a regex and how those indices are aligned.